text/0000-http-index.md

```diff
@@ -0,0 +1,139 @@
+- Feature Name: http_index
+- Start Date: 2019-10-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Selective download of the crates-io index over HTTP, similar to a solution used by Ruby's Bundler. Changes transport from an ahead-of-time Git clone to HTTP fetch as-needed, while keeping existing content and structure of the index. Most importantly, the proposed solution works with static files and doesn't require custom server-side APIs.
+
+# Motivation
+[motivation]: #motivation
+
+The full crate index is relatively big and slow to download. It will keep growing as crates.io grows, making the problem worse. The requirement to download the full index slows down the first use of Cargo. It's especially slow and wasteful in stateless CI environments, which download the full index, use only a tiny fraction of it, and throw it away. Caching of the index in hosted CI environments is difficult (`.cargo` dir is large) and often not effective (e.g. upload and download of large caches in Travis CI is almost as slow as a fresh index download).
+
+The kind of data stored in the index is not a good fit for the git protocol. The index content (as of eb037b4863) takes 176MiB as an uncompressed tarball, 16MiB with `gz -1`, and 10MiB compressed with `xz -6`. Git clone reports downloading 215MiB. That's more than just the uncompressed latest index content, and over **20 times more** than a compressed tarball.
+
+A while ago, GitHub indicated they [don't want to support shallow clones of large repositories](http://blog.cocoapods.org/Master-Spec-Repo-Rate-Limiting-Post-Mortem/). libgit2 doesn't support shallow clones yet. Squashing of the index history adds complexity to management and consumption of the index (which is also used by tools other than Cargo), and still doesn't solve problems of the git protocol inefficiency and overall growth.
```

Comment:
> A while ago, GitHub indicated they don't want to support shallow clones of large repositories.

Having clicked through the links[ the comments in the original issue](https://github.com/CocoaPods/CocoaPods/issues/4989#issuecomment-193838934) that triggered the blog post seem more nuanced.

E.g. 

> To reiterate: the major performance issue is not on doing an initial clone of the CocoaPods repository, but in performing a shallow clone and then repeatedly fetching into it, like the CocoaPods client is currently doing.

It sounds like shallow clones themselves are not as much of a problem as repeated fetches. Maybe that part could be clarified or removed.

That said, libgit2 [still doesn't support it](https://github.com/libgit2/libgit2/issues/3058), so the other points are still valid.

Reply:
Ah, but this may enable an alternative approach. If libgit2 doesn't support it then cargo could opportunistically try to use the system git to perform a shallow clone. It should be available in many CI environments already.

Reply:
Cargo does fetch into the repository regularly, so it can't start using shallow clones, because it would create the exact problem described in that issue: "**performing a shallow clone and then repeatedly fetching into it**".

Reply:
Yes, for local development copies that makes sense. But isn't one justification for HTTP fetches un-cached CI? In that case shallow copy would make sense.

But yeah, I guess it's not a good alternative anyway.

Reply:
Oh I see. A one-off shallow copy that is never updated again would be a bit more efficient than clone and acceptable from GitHub's perspective. However, it would still download more data than the sparse HTTP fetch. That's because even a shallow clone still fetches all data for all crates.